### PR TITLE
add OwnedEntityPicker field in plugin/scaffolder 

### PR DESCRIPTION
**Pull request by mufaddal7**
## Hey, I just made a Pull Request!

Added a new field component in `plugin/scaffolder`

`OwnedEntityPicker` lists down all Entities owned by the logged-in user 

```yaml
spec:
  parameters:
      properties:
        EntityOwner:
          title: Your Entities
          type: string
          description: List of Entities owned by you
          ui:field: OwnedEntityPicker
          ui:options:
            allowedKinds:
              - System
            defaultKind: System
```
For testing the field use may use this [template](https://gitlab.com/mufaddalWH/express-app/-/blob/master/templateEntityOwners.yaml) 

https://github.com/backstage/backstage/issues/8391

### CODE REVIEW

**Comment by benjdlambert** [comment]
Thanks for this! This is a great start!

Thinking that this is a little inefficient at the min, and I know that it's done this way in a a lot of other places too, with how we load all entities and then use the filter to check if we own it or not.

I'm wondering if we can have a go at trying doing this part from the hook https://github.com/backstage/backstage/blob/31f00e8015ef5eabbc586a5ef28e4e4871b61194/plugins/catalog-react/src/hooks/useEntityOwnership.ts#L145-L146 and building a `Set` of all the ownership refs, and then make one query `getEntities` and using the `filter` prop to specify `ownedBy = ref1` `OR` `ownedBy = ref2`

**Comment by benjdlambert** [comment]
So I worked it out because I couldn't find an example anywhere of doing this query.

But you want to grab the `identityRefs` and `catalogRefs` like before and pass them as the array to like below:

```ts
const catalogApi = useApi(catalogApiRef);
const response = useAsync(() =>
    catalogApi.getEntities({
      filter: {
        [`relations.${RELATION_OWNED_BY}`]: ['group:default/team-b', 'user:default/guest'],
      },
    }),
  );
  ```

Maybe it makes sense actually to make a hook in `catalog-react` that encapsulates all this and use the new hook instead of the hook that returns you a matching function. `useOwnedEntities` or something.

**Comment by mufaddal7** [comment]
Yes, that can be a good addition as well. 
I spend some time finding an API that could do the specific query to get owned entities, but couldn't find anything. Hence went with this approach

**Comment by benjdlambert** [comment]
Yep - I tried looking for examples too, and it was a pain so had a little play to see if I could get it working. Thinking that we will probably want to use this approach more than the other to be honest where we can as it's really not performant for large catalogs to grab the entire catalog and filter in the client. Much better to do it serverside where we can cache the filtering etc too if needed to increase performance later down the line.

**Comment by regicsolutions** [comment]
Would this restrict a user from inputing a entity name manually? Also, what happens when the user ownes 0 entities?

**Comment by mufaddal7** [comment]
@benjdlambert the code snippets you shared were very helpful, created a `useOwnedEntities` file, have placed it in the same folder, for now, maybe later we can push it to `@backstage/catalog-client` 

Now our component will fetch only the Entities owned by the user. Please review

**Comment by mufaddal7** [comment]
> but was wondering if this would restrict the current user from inputing a entity name manually
As this is only a front-end change, I don't think it will restrict or validate user's input, you might want to use custom `FieldValidation` for this.

> We have a use case that came up around the Group entity where we were looking to have the user only pick a Group they are a member of
We can use this component and pass groups in `allowedKinds`

> Also, what happens when the user ownes 0 entities?
No options will be visible to the user

File: plugins/scaffolder/src/components/fields/OwnedEntityPicker/useOwnedEntities.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  catalogApiRef,
+  loadCatalogOwnerRefs,
+  loadIdentityOwnerRefs,
+} from '@backstage/plugin-catalog-react';
+import { identityApiRef, useApi } from '@backstage/core-plugin-api';
+import { Entity, RELATION_OWNED_BY } from '@backstage/catalog-model';
+import { CatalogListResponse } from '@backstage/catalog-client';
+import { useAsync } from 'react-use';
+import { useMemo } from 'react';
+
```
Comment (benjdlambert) [inline plugins/scaffolder/src/components/fields/OwnedEntityPicker/useOwnedEntities.ts:26]: I think this hook should be in `catalog-react` package instead.

Comment (mufaddal7) [reply plugins/scaffolder/src/components/fields/OwnedEntityPicker/useOwnedEntities.ts]: Did the change. It's now in `catalog-react`

File: .changeset/eleven-baboons-sparkle.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-catalog-react': minor
```
Comment (benjdlambert) [inline .changeset/eleven-baboons-sparkle.md:2]: Both changeset's should be `patch` not `minor` 👍

Comment (mufaddal7) [reply .changeset/eleven-baboons-sparkle.md]: Changed it 👍

**Comment by regicsolutions** [comment]
Thanks @mufaddal7 this helps! It looks like there was a recent contribution to allow an option to disable arbitrary values within the EntityPicker - https://github.com/backstage/backstage/pull/8441 do you feel it makes sense to also include this option as well @benjdlambert ?

**Comment by benjdlambert** [comment]
Let's leave that for a follow up PR - want to try and get this in for the release today :pray:

**Comment by mufaddal7** [comment]
All checks cleared , feels good. 
No problem,  next week it is 🎉 @benjdlambert

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**